R/trace_tests.R

```diff
@@ -97,17 +97,60 @@ NULL
 count_test <- function(key) {
   n_calls_into_covr <- 2L
 
-  if (is_current_test_finished())
-    update_current_test(key)
+  if (is_current_test_finished()) {
+    update_current_test()
+  }
 
   # ignore if .counter was not created with record_tests (nested coverage calls)
   if (is.null(.counters[[key]]$tests)) return()
 
-  depth_into_pkg <- length(sys.calls()) - length(.current_test$frames) - n_calls_into_covr + 1L
   .current_test$i <- .current_test$i + 1L
-  .counters[[key]]$tests <- rbind(
-    .counters[[key]]$tests,
-    c(length(.counters$tests), depth_into_pkg, .current_test$i)
+
+  # expand infrequently as new tests are added, doubling matrix size as needed
+  tests <- .counters[[key]]$tests
+  n <- NROW(tests$tally)
+  if (.counters[[key]]$value > n) {
+    tests$tally <- rbind(tests$tally, matrix(NA_integer_, ncol = 3L, nrow = n))
+  }
```

Comment:
Repeated calls to `rbind` to add a single row to a matrix were causing substantial slowdown due to reallocation with each `rbind`. Instead, the matrix now starts with 4 rows and doubles in size each time it is filled. When merging all the traces together, any unused rows are filtered out (see `clean_coveage_tests` & `merge_coverage_tests`)

Reply:
Yeah, repeated calls to rbind is basically the worst case performance pitfall in R.

Did you try storing the intermediate results in a list and then combining them in one pass with `do.call(rbind.list)` or faster equivalents like `rlang::vec_rbind()`? Generally those types of solutions end up being the fastest solution to this problem.

Reply:
Thanks @jimhester - I'll try that method out and report back. That would save some of the complexity of dealing with partially used matrices.

Reply:
@jimhester - I tried out a few approaches using `list`s or `environment`s to collect individual rows and then collecting them into a single matrix using `do.call(rbind, <list>)`. I wasn't able to outperform this implementation. I didn't try using the `rlang` alternatives because I don't think this behavior warrants the introduction of a new dependency. 

Assigning into a `.counters[[key]]$tests` `list` (eg, `.counters[[key]]$tests[[n]] <- val`) hit the same performance issues with nested listed assignment that I was running up against before. To get comparable performance, I still had to wrap it in an environment to avoid nested assignment which means the approach still has the same complexity in how the data is structured. 

Overall, I think the approach in the initial PR is probably still the best option.